Fix fs.utimesSync issue and document touchHits option

mseminatore opened the pull request:
@avik-das For review per our discussion.

avik-das commented:
@mseminatore: Is this happening on any usage of sepia, or in a particular scenario? All the automated tests pass for me on node.js v6.9.1 even without your change. I wonder if this is Windows-only.

If there's a specific scenario when this problem occurs, can you add an automated test, whether it's a unit test or one in the `examples` directory, that fails before your change and passes after? If this is happening in all cases, then the existing tests should be failing for you before your change. Can you confirm that?

Also, after your change, one of the unit tests is failing. You need to make a small tweak to the test. Are you running the automated tests before putting up the PR?

avik-das commented on src/util.js:
```diff
@@ -177,7 +177,7 @@ function touchOnHit(filename) {
 
   filename = filename + '.headers';
-  var now = Date.now();
+  var now = Infinity;	// per NodeJS this will correctly set timestamp to Date.now() / 1000
```
@mseminatore: Minor: you're using a tab for spacing, when you should be using spaces. Also, this line is now over 80 characters wide, when the style for this codebase is 80 characters. I realized that I need to add a jshint setting for that, and I'll make sure to do that.

mseminatore replied:
Sorry about tab vs. spaces.  I expected that the comment would be removed.

mseminatore commented:
The tests fail in numerous places on Windows.  For example, in test/util.js you compare a constructed path to a string literal path in many places resulting in errors when 'd:\global\fixture\...' is invalidly compared to '/global/fixture/...'  Because of the scope of the test issues I was unable to use them and instead proposed a targeted one line fix proposal.

If you refer to the NodeJS docs you will see that Date.now() is an invalid value to pass to fs.uTimesSync() since the former is in milliseconds and the later takes Unix timestamps which are seconds.  Whether or why this would work in your test cases is not clear to me without testing on a different OS.

avik-das commented:
@mseminatore: Thanks for the clarifications. This helps me understand both the scope of the issue, as well as what the fix is attempting to do. I also did some more digging, and it turns out that on Linux, you can simply set the timestamp in the future! Your example code actually succeeds on my machine, but it sets the timestamp in the future, and dividing `Date.now()` by 1000 first sets it to the current time.

So, the correct fix here is to use `Date.now() / 1000` instead of `Infinity` because it's clearer. I also fixed the unit test accordingly. I went ahead and redid the commit myself, but I brought in your documentation change, so I added you to the contributors list.

Finally, I published version 2.0.2 of sepia, so you should be able to pull it in and be unblocked. Thanks for digging up this issue and explaining it. In the future, explaining why you couldn't use the automated tests in the first place would help avoid confusion. I'm going to close this PR.

mseminatore commented:
Certainly!  I had originally used Date.now() / 1000 as my fix.  However in
reading the Node docs had noted that Infinity and NaN signaled now.

That said I agree that it is generally clearer to use Date.now() / 1000
than a special case that could change in the future.  I've updated to 2.02
and it works great thank you!


On Mon, Dec 5, 2016 at 10:53 PM, Avik Das <notifications@github.com> wrote:

> @mseminatore <https://github.com/mseminatore>: Thanks for the
> clarifications. This helps me understand both the scope of the issue, as
> well as what the fix is attempting to do. I also did some more digging, and
> it turns out that on Linux, you can simply set the timestamp in the future!
> Your example code actually succeeds on my machine, but it sets the
> timestamp in the future, and dividing Date.now() by 1000 first sets it to
> the current time.
>
> So, the correct fix here is to use Date.now() / 1000 instead of Infinity
> because it's clearer. I also fixed the unit test accordingly. I went ahead
> and redid the commit myself, but I brought in your documentation change, so
> I added you to the contributors list.
>
> Finally, I published version 2.0.2 of sepia, so you should be able to pull
> it in and be unblocked. Thanks for digging up this issue and explaining it.
> In the future, explaining why you couldn't use the automated tests in the
> first place would help avoid confusion. I'm going to close this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/avik-das/sepia/pull/1#issuecomment-265075296>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQDQuE1vwg8OX0T-Wv0ouq9jhZYUfoc5ks5rFQZ0gaJpZM4LCukr>
> .
>
